### PR TITLE
Add "author data" functionality to SolrDefault

### PR DESCRIPTION
**Pull request by datavoyager**
This patch will enable users to append author related data besides roles
(e.g. Internal identifiers, ORCIDs etc) to the deduplicate author array

### CODE REVIEW

**Comment by demiankatz** [comment]
Thanks for this! A few comments;

1.) Is the authorFields property actually appropriate/necessary? Given that all of the deduplication logic related to main, secondary and corporate is still hard-coded into getDeduplicatedAuthors(), I'm wondering if it would be more appropriate to just have the array defined in that method as well. Defining it as a property seems to imply that you can put arbitrary values in there and still have the code work, but as currently coded, taking away any of the expected values or adding new ones actually won't work correctly (and I suspect that actually coding the deduplication logic to be fully generic is more work than it's worth if your use case doesn't require it).

2.) Did you consider writing this so that it was generically constructing function names rather than Solr field names? As currently written, we potentially lose a level of abstraction, and by going to the raw Solr data instead of using the usual method calls, we might introduce subtle problems if people have layered extra processing into the get methods. At a glance, it seems to me that you could probably code something with tryMethod() and constructed method names that's practically the same as what you're doing here with constructed field names.

3.) If you agree with me about switching from Solr fields to method names, this comment may become irrelevant... but if you think we need to keep the getField() method in here, I think it should be made protected rather than public -- we don't really want to encourage external logic to fetch out raw Solr data unless absolutely necessary. On a side note, a method like getField may be generally useful to make the whole driver shorter and slightly more readable -- I've toyed with the idea of adding it just to simplify all the existing getters. Not sure whether or not it's worth the extra function call overhead it introduces, though, because a single ternary call really isn't that hard to read. Might be preferable to just leave it alone -- at some point in the future we'll be able to raise the minimum PHP version to 7, at which point we can use the new [?? operator](http://lornajane.net/posts/2015/new-in-php-7-null-coalesce-operator) to simplify tons of code in an even more readable way.

4.) This seems to introduce a lot of complex, repeated template logic. Did you consider the possibility of a sub-template or view helper method to simplify this and improve readability?

I'd also be interested to hear what @lahmann thinks of this if he has a chance to take a look.

**Comment by datavoyager** [comment]
Hi Demian,

Thanks for your comments. Here are some replies:

1) I had thought about advancing this patch further to include an "authorFields" setting loaded from configuration. That would increase the flexibility of the functionality to users could add their own author fields. e.g. author_institutional etc I'm not sure if that's taking things too far however as the existing author fields are obviously standard. Perhaps a compromise would be to hardcode author, author2 and corporate (as it currently does) and add a method which will also add any custom user authors?

2) + 3) I don't have any objections to switching to a method name based approach.

4) I'll add some templating / view helper logic as you suggest

Thanks,
Luke

**Comment by demiankatz** [comment]
1) I certainly don't object to the idea of having support for arbitrary author fields -- I'm just concerned that if we add that support, we need to be careful about making sure the deduplication logic behaves correctly and that the algorithm scales correctly to N author fields. Because that seems like it might end up being a lot of work and adding significant complexity, my gut feeling is we're better off waiting until we need it before we invest effort into implementing it... but if you think it's simpler than I'm anticipating, or if you know you're going to need it, I'm definitely open to addressing the problem now (or, if you prefer, attacking that with a separate pull request after we've gotten the other baseline issues sorted out here).

2/3/4) Thanks -- please comment again when you're done and I'll give it another look!

**Comment by datavoyager** [comment]
Hi Demian,

I've completed some more refactoring. I've removed all the protected variables and created a View Helper method and corresponding template.

**Comment by demiankatz** [comment]
This definitely looks like a step in the right direction to me!

A few more quick thoughts:

1.) This looks like a significant backward-compatibility break; I wonder if we should delay merging this until 4.0. If you're in a hurry to have this in master, though, we can look into mitigating the changes... I'm just concerned that this could potentially break a lot of local custom templates. (That's not a reason not to do it -- but it just means we need to pick our timing more carefully).

2.) There are a few style issues being picked up by Travis. Can you fix these, or would you like me to help?

3.) Now that we have a view helper method for the rendering of the author list, did you consider the possibility of moving more of the business logic into the method and out of the template? It may or may not be feasible to pre-process the data to allow for simpler templating... but if it is, it's probably worth doing. If you feel it's less complicated as-is, though, we certainly don't have to do further refactoring here.

Thanks!

**Comment by datavoyager** [comment]
Hi Demian,

Here are some replies:

I think I'm right in saying that the only backward compatibility issue is that I have renamed the "main" author index as "primary". I can change this back to "main" though as I manually pass the value which checks for datafields into the corresponding method.

e.g.   $authors['main'] = $this->getAuthorDataFields(
            "primary", $this->getPrimaryAuthors(), $dataFields
        );
2) I'll look into the styling issues
3) I'll look for more refactoring opportunities

Thanks,.

Luke

**Comment by demiankatz** [comment]
Regarding backward compatibility, I think the other thing is that you've added another layer of data in order to accommodate extra fields beyond roles. This absolutely makes sense and seems like the right thing to do, but it will break legacy custom templates since they're assuming an author => roles associative array. I can't really think of an obvious, straightforward way around that.

**Comment by demiankatz** [comment]
@datavoyager, now that we're beginning work on release 4.0 and backward compatibility breaks are not such a problem, we can revisit this. I just did a quick review and had a few comments:

1.) It looks like in the most recent merge, some conflict markers got committed in core.phtml. We definitely need to sort that out before proceeding.

2.) On a possibly related note, it seems like the refactoring to the deduplicated-authors.phtml template has caused the microdata markers (e.g. `<span property="creator">`) to get dropped from core.phtml.

3.) Some of this work now overlaps with #829, which refactors more of the data fields to separate sub-templates. Perhaps it is best to put this on hold until #829 is resolved, since that may simplify some of the implementation details here. I'm interested in your opinion on the #829 refactoring if you have time to look at it.

Thanks!

**Comment by demiankatz** [comment]
@datavoyager, now that #829 has been merged, I think this needs to be reimplemented. I'll see if I can find time to adapt this to work with the new mechanism and open up a replacement PR for your review. If you think you're going to have time for this yourself, let me know and I'll happily defer to you -- but otherwise, I don't mind doing it myself as soon as time permits.

**Comment by datavoyager** [comment]
@demiankatz  - Thanks. I'll take a look at it after Christmas :)

**Comment by demiankatz** [comment]
@datavoyager, much appreciated. Enjoy Christmas and New Year's, and we can regroup in 2017 -- as always, let me know if I can be of any assistance! (And thanks for your willingness to work on this -- while I'm happy to do the work if necessary, I think it will be a good test of the new architecture to be sure that somebody else can work with it comfortably).

**Comment by demiankatz** [comment]
@datavoyager, just checking whether this is still on your radar -- guessing it may have been lost in the holiday shuffle. If you need me to take a look on this end, please let me know! (No rush, in any case -- just making sure we don't forget about it entirely).

**Comment by datavoyager** [comment]
@demiankatz sorry! This did fall through the holidays cracks. I've created a ticket to review it as part of my workflow to remind me to look at it.

**Comment by demiankatz** [comment]
Great, thanks -- I'm happy to help if there's anything I can do once you've had a chance to look things over.

**Comment by datavoyager** [comment]
Hi @demiankatz 

I've committed some changes to the branch to tie it into the RecordDataFormatter Code.

I think the main things you may want to review is the method and variables names (I've used dataField as the base for values like roles and tried to apply that consistently from RecordDriver to Formatter to Template).

You may also want to review the complexity of data-authors.phtml - personally, I don't think it's too bad but it does introduce extra foreach and if statements.

Thanks,

Luke

**Comment by demiankatz** [comment]
Thanks, @datavoyager -- at a glance this looks good to me. I'll try to run some tests and give it a more detailed look over the next week or two.

File: themes/bootstrap3/templates/RecordDriver/SolrDefault/data-authors.phtml [hunk 1]
```diff
@@ -1,15 +1,29 @@
...
     };
-    return ' (' . implode(', ', array_unique(array_map($translate, $roles))) . ')';
-  };
+    return '<span class="author-property-' . $name . '">(' . implode(', ', array_unique(array_map($translate, $datafield))) . ')</span>';
+};
 ?>
 <? if (isset($data[$type]) && !empty($data[$type])): ?>
-  <? $i = 0; foreach ($data[$type] as $author => $roles): ?><?=($i++ == 0)?'':', '?><span property="<?=$this->escapeHtml($schemaLabel)?>"><a href="<?=$this->record($this->driver)->getLink('author', $author)?>"><?=$this->escapeHtml($author)?></a><?=$formatRoles($roles)?></span><? endforeach; ?>
+  <? $i = 0; foreach ($data[$type] as $author => $dataFields): ?>
+    <?=($i++ == 0)?'':', '?><div property="<?=$this->escapeHtml($schemaLabel)?>">
```
Comment (demiankatz) [inline themes/bootstrap3/templates/RecordDriver/SolrDefault/data-authors.phtml]: There seem to be some layout problems here -- the reason the old code was all jammed into a long, hard-to-read line is that if you introduce whitespace, it causes spaces to appear in front of commas, which is not desirable. Also, it looks like some spans have been changed to divs, which is causing unwanted newlines in the list.

**Comment by datavoyager** [comment]
@demiankatz Thanks for looking at this. I'm happy with the changes you have made regarding  $author[$i] / $role[$i] - that was an oversight on my part.

**Comment by demiankatz** [comment]
Thanks for the review, @datavoyager. I've clarified the code as you suggested and also added some comments for further readability. I think that just leaves the template issue outstanding. I can look at that if you don't have time -- just let me know if you need me to!

**Comment by datavoyager** [comment]
@demiankatz I've had a stab at the template issues - basically a combination of css to inline the author details div and output buffering to capture and trim the content.

I also added a check for empty data fields to prevent them being added

**Comment by demiankatz** [comment]
@datavoyager, I still had some problems with comma whitespace, but I've pushed up some adjustments that seem to have that issue under control. This PR also brought to my attention the fact that the collection-*.phtml files are not using the recordDataFormatter (an oversight on my part). I think I should clean that up and merge the changes in here before proceeding, since that will make the code over here somewhat simpler. Once I have that done, I'll do a bit more testing and let you know if any issues remain. Thanks again for the help -- more soon!

**Comment by datavoyager** [comment]
@demiankatz Thanks Demian :)

**Comment by demiankatz** [comment]
@datavoyager, as promised, I've made improvements to the use of RecordDataFormatter in master and merged them here, which centralizes all of the author display logic to just one template. Next week, I'll try to find some time to make a few more adjustments (adding comments, possibly simplifying some code) and will ask @crhallberg to take one last look at the CSS and template. Once all that's done, we'll get this merged to master. Thanks again for all of your help!

**Comment by crhallberg** [comment]
The code looks good overall. I'm only wondering why you're using a div instead of a span, especially since it's being set to `display: inline`.

**Comment by datavoyager** [comment]
@crhallberg If author datafields are set (e.g. role), the div can hold a number of span elements.

**Comment by demiankatz** [comment]
@datavoyager, is there a reason that the div can do this but a span can't? I thought that the primary difference between the two was block vs. inline behavior.... Sorry if I'm just confused, though. :-)

**Comment by datavoyager** [comment]
@demiankatz @crhallberg Sorry! I though you weren't supposed to nest inline elements within each other but I've checked the html5 specs and you can.

**Comment by demiankatz** [comment]
Okay, great, that explains what was going on there. I think that means we can change the div back to a span and remove the custom inline styling, which will simplify this PR a little bit more! I'll take care of that shortly.

**Comment by demiankatz** [comment]
@datavoyager, I've just merged the $default parameter on tryMethod into master, since that's a useful stand-alone improvement. That just leaves three changed files here, which I suspect can be merged very soon -- I just want to review them again with fresh eyes later in the week.

**Comment by demiankatz** [comment]
@datavoyager, sorry my final review took so long, but I finally got around to it. Merging now. Thanks again!!

**Comment by datavoyager** [comment]
Thanks for merging it and for all your help @damiankatz :)

**Pull request merged**